### PR TITLE
Iterdataset

### PR DESCRIPTION
**Pull request by tung-nd**
This avoids loading all data into RAM in the beginning, but we'll have to trade off randomness. The idea is first to chunk the netcdf data into smaller npz files, and then load these files one by one. We shuffle these files and have to buffer to have some level of randomness.

Can use the local notebook test.ipynb for testing. It works well for Forecasting, will have to make some changes for downscaling.

### CODE REVIEW

**Comment by tung-nd** [comment]
@jasonjewik @prakhar6sharma feel free to take a look when you have the time

**Comment by tung-nd** [comment]
TODO:
- [ ] test on downscaling
- [ ] add time information (needed for visualization)

**Comment by tung-nd** [comment]
Forgot to mention, the command to preprocess and chunk the netcdf data into npz files is
```
python src/climate_learn/data/nc2npz.py \
--root_dir /data0/datasets/weatherbench/data/weatherbench/era5/5.625deg/ \
--save_dir /data0/datasets/weatherbench/data/weatherbench/era5/5.625deg_npz \
--start_train_year 1979 --start_val_year 2016 --start_test_year 2017 --end_year 2018 \
--num_shards 16
```
I've already run this. The preprocessed data is in `/data0/datasets/weatherbench/data/weatherbench/era5/5.625deg_npz`

File: src/climate_learn/models/components/resnet.py [hunk 1]
```diff
@@ -1,8 +1,7 @@
-from math import log
 import torch
 from torch import nn
 from torch.distributions.normal import Normal
-from .cnn_blocks import PeriodicConv2D, ResidualBlock, Upsample
```
Comment (jasonjewik) [inline src/climate_learn/models/components/resnet.py:5]: Why was `Upsample` removed?

File: src/climate_learn/models/components/unet.py [hunk 2]
```diff
@@ -41,7 +40,6 @@ def __init__(
             out_channels = in_channels
         self.out_channels = out_channels
         self.hidden_channels = hidden_channels
-        self.upsampling = upsampling
```
Comment (jasonjewik) [inline src/climate_learn/models/components/unet.py:44]: Why was upsampling removed?

File: src/climate_learn/models/components/vit.py [hunk 3]
```diff
@@ -30,7 +30,6 @@ def __init__(
             "10m_v_component_of_wind",
         ],
         out_vars=["2m_temperature"],
-        upsampling=1,
```
Comment (jasonjewik) [inline src/climate_learn/models/components/vit.py:33]: Why was upsampling removed?

**Review by jasonjewik — COMMENTED** [review]
Can I get clarification on why upsampling was removed from the models? I'm still working through the data stuff.

**Comment by tung-nd** [comment]
Previously we perform downscaling by having a few upsampling layers at the end. However, a better way to do it is to first bilinearly interpolate the input to match the output dimensions, then feed the interpolated input to the network. I found this work better in my previous experiments, and this avoids the need to know the upsampling ratio when constructing the network.

**Review by jasonjewik — CHANGES_REQUESTED** [review]
@tung-nd I tested this code on MINT2. Can you confirm `test.ipynb` works there? If it works, then delete `test.ipynb` and commit again.

**Review by jasonjewik — APPROVED** [review]
